### DISCUSSION TITLE
NixOS guest?

### DISCUSSION THREAD

**Discussion by LATBauerdick** [lima-vm/lima#430] [Q&A] [answered]
I would really love to have a NixOS guest -- at the moment I use lima with the nix package manager on an archlinux guest. I think in theory NixOS should work, looking at the stated requirements. I also have a qemu NixOS VM, that works just fine. It does not have the great lima integration features, so I really want it running from lima.

I guess the best way forward would be to just go ahead and do it myself, but I am new to lima and could not (easily) find a thread how to start. Are there some example instructions somewhere, or where should I look to e.g. how it was done for archlinux or Debian?

Any pointers are much appreciated.

**Comment by jandubois** [comment]
> I guess the best way forward would be to just go ahead and do it myself

Yes, it would! 😄 

The first step is to find a somewhat stable/official location to download cloud images from.

The second step is implementing support for the package manager in `pkg/cidata/cidata.TEMPLATE.d/boot/30-install-packages.sh`.

The third step is trying it out, checking if something is broken, and fixing it. 😈

**Reply by ghost** [reply]
I am also interested in a NixOS guest and may be able to provide some development.

I did try this with the minimal 64-bit iso image I found at [NixOS' download](https://nixos.org/download.html).

It did not work correctly and I did not see anything in `serial.log`.

From what I researched NixOS itself has a way to build isos. We could leverage that to build a pared down image with the appropriate configuration to integrate with lima.
This comment gave me that idea: https://discourse.nixos.org/t/how-to-build-a-standalone-nixos-qemu-vm/5688/3

So we could try step 2 and 3 of what @jandubois notes!

**Comment by LATBauerdick** [comment]
great! I think I already roughly know how to do the first step, to create a working qemu image and install additional stuff, via a script, starting from the official/current NixOS .iso image (pulling it from the NixOS archive). 

I will put this together over the holidays and report back.

I am however pretty hazy about the next steps, and will have to look at the repository in detail to see how it works with alpine or arch linux and go from there.

I am far from an expert though, help will be much appreciated!

**Comment by jandubois** [comment]
Just a quick note that there are 2 different scenarios: booting from an image vs. booting from an ISO. We latter is only done for Alpine right now, and we build a custom ISO to make sure it includes cloud-init (or our own lima-init replacement). Most ISOs are either install or rescue media and do not support this.

There is special code in Lima to mount a persistent data volume and move some directories to it, so they survive a reboot. This code probably will need some adjustments if you want to boot NixOS from an ISO instead of a regular image.

Let me know if you run into any issues, and I'll try to help out!

**Comment by afbjorklund** [comment]
Apparently NixOS thinks that `cloud-init` is too bloated to go into the default image, but it has been packaged:

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/system/cloud-init.nix

* Issue https://github.com/NixOS/nixpkgs/issues/39076

I don't think it is the python3 itself that is the biggest bloat, but it (ci) does have a truckload of dependencies...

**Reply by jandubois** [reply]
> Apparently NixOS thinks that `cloud-init` is too bloated to go into the default image, but it has been packaged:

If you build you own image, you could also use the `lima-init` I wrote for our Alpine ISO instead. I it just a shell script and some `awk` code, and supports everything needed by Lima from cloud-init. It may need some extra work to make it compatible with nix.

**Reply by afbjorklund** [reply]
You have to fork cloud-init as well, since the list of distros is hardcoded within... So it might be easier to start with lima-init

**Comment by vielmetti** [comment]
Also registering interest in this development!

**Comment by patryk4815** [comment]
Today i created my nixos image https://github.com/patryk4815/ctftools/tree/master/lima-vm
And it is working with limactl :)

**Reply by stuta** [reply]
From where do I get nixos-aarch64.img?

**Reply by patryk4815** [reply]
@stuta you need build it localy, just run this command and you will get image :)
`nix build .#packages.aarch64-linux.box`

**Reply by stuta** [reply]
I'm missing some context here. So I installed default Ubuntu to OSX M1 Lima and then installed Nix to Ubuntu with ` sh <(curl -L https://nixos.org/nix/install) --daemon`.

Then I did go to lima-vm -folder and ran: `nix --extra-experimental-features nix-command --extra-experimental-features flakes build .#packages.aarch64-linux.box`

After a long download I get an error `
error: a 'aarch64-linux' with features {kvm} is required to build '/nix/store/yxy5233kbb26kv3i18ylq7mczm6nf8cw-nixos-disk-image.drv', but I am a 'aarch64-linux' with features {benchmark, big-parallel, nixos-test}`.

**Reply by patryk4815** [reply]
ah yes, you need "kvm" feature. But anyway you can "bypass" this check.
Put this into `/etc/nix/nix.conf`:
```
system-features = nixos-test benchmark big-parallel kvm
```

(qemu in this case will use TCG mode not kvm)

**Reply by stuta** [reply]
I had to do limactl stop & start for this setting in /etc/nix/nix.conf to work. Now it built the disk image.

**Reply by stuta** [reply]
Just to clarify for others: running `nix build .#packages.aarch64-linux.box` created a symlink called result (/nix/store/rhg5kv530fg93ylhncywmpsfc4x01hqw-nixos-disk-image) and inside that folder is a new disk image. 

Move image from lima to OSX lima-vm/imgs -folder with name nixos-aarch64.img. Then fix aarch64.yml (line 2 has a space, remove it and fix memory and others). Then move OSX ~/.lima/default to somewhere else and run:
```bash
limactl delete -f default
limactl start --name=default aarch64.yml
lima
```

Inside lima run:
```bash
uname -a
Linux nixos 5.17.3 #1-NixOS SMP Wed Apr 13 17:27:43 UTC 2022 aarch64 GNU/Linux
```

Thanks!

**Reply by msgilligan** [reply]
@patryk4815 's tools with @stuta 's clarifications worked for me. I used a Raspberry Pi 5 running Pi OS (Debian) to build the image then transferred it to my Mac and it worked!

**Reply by afbjorklund** [reply]
Is there anything more needed in lima, for this to work ? Seems like it is working now, with the lima.yaml and the .img provided

**Reply by patryk4815** [reply]
@afbjorklund My "image" does not use cloud-init, I manually rewrote cloud-init to ".nix module" and in runtime/everyboot fetch data from the file "/mnt/lima-cidata/user-data" and "/mnt/lima-cidata/lima.env".
Do you want to support this image upstream?

**Reply by afbjorklund** [reply]
Not the image (nor the yaml), the question was if anything was needing changes in lima for it to work - but it seems not?

**Reply by afbjorklund** [reply]
Rewriting cloud-init is "optional", I suppose. The alpine-lima and tinycloud images also use non-official/non-python parsers

**Reply by jandubois** [reply]
> Do you want to support this image upstream?

I think this would only be possible if the images were either published by the Nix org itself, or alternatively build by a `lima-vm` org repo and published by GitHub action to `lima-vm` GitHub releases. And then it would still require a Lima maintainer to agree to provide ongoing maintenance.

Right now we only have a single image (the "alpine" ISO one) that is being maintained by Lima maintainers, and I suspect it is unlikely that we will add more.

**Reply by msgilligan** [reply]
@patryk4815 wrote:

> My "image" does not use cloud-init, I manually rewrote cloud-init to ".nix module" and in runtime/everyboot fetch data from the file "/mnt/lima-cidata/user-data" and "/mnt/lima-cidata/lima.env".

Why did you choose to do it this way rather than directly use cloud-init?

**Reply by msgilligan** [reply]
> > Do you want to support this image upstream?
> 
> I think this would only be possible if the images were either published by the Nix org itself, or alternatively build by a `lima-vm` org repo and published by GitHub action to `lima-vm` GitHub releases. And then it would still require a Lima maintainer to agree to provide ongoing maintenance.

The ability to run and configure NixOS in Lima VMs is important to me. I just got @patryk4815's solution working and need to test it out a little more and learn more about how it works. I would potentially be interested in helping to make this a third-party add-on for Lima (since it is unlikely we'll be able to get it published by Nix.org or by the Lima maintainers.)

The first steps would be giving it a dedicated repository with its own top-level README, adding documentation and making minor improvements.  I'm not sure if it makes sense to put something like this in nixpkgs, but that could also be an option.

I've already made one change to the `aarch64.yml` config file, I added: 

```
ssh:
  forwardAgent: true
```

which works well.  (And is one of my favorite Lima features)

This week I plan on experimenting with the NixOS VM that I build (and perhaps rebuilding it, etc.) by using it to work on some of my projects and seeing if I can use it to build another NixOS/Lima VM, etc.

I would also like to understand the best ways to customize the configuration of NixOS/Lima VMs. There are at least three approaches I can think of:

1.  Modify or override the various `.nix` files in the `ctftools/lima-vm` directory before the `nixos.img` is generated.
2. Use some (ideally Nix-based) mechanism to customize the VM during the boot/provisioning phase of Lima
3. Use NixOS generation in the VM after it is started

One example of something I would like to change/customize is to add the "nix-command" and "flakes" features automatically. I would also like to be able to build VMs with particular services installed, etc.

**Reply by msgilligan** [reply]
@patryk4815 Are you interested in improving or maintaining your solution? I find it useful enough to want to spend at least some time moving it forward. Should I just create a new repository and give you credit, or do you (or anyone else) want to participate?

**Reply by sam-goodwin** [reply]
The img is 3.2G, is that expected? The official NixOS ISO is only 1GB.

**Reply by msgilligan** [reply]
> The img is 3.2G, is that expected? The official NixOS ISO is only 1GB.

The one I built was 3250585600 bytes.

**Comment by msgilligan** [comment]
I have found two projects for building Lima-compatible NixOS images on GitHub that are inspired by @patryk4815 's work:

* https://github.com/kasuboski/nixos-lima ("Heavily inspired from patryk4815/ctftools") Blog post: [NixOS Dev Environment on Mac](https://www.joshkasuboski.com/posts/nix-dev-environment) 
* https://github.com/quinneden/nixos-lima-vm ("Forked from kasuboski/nixos-lima, still mostly identical, slight modification in the user configuration and lima runtime.")

**Reply by sam-goodwin** [reply]
https://github.com/kasuboski/nixos-lima - this one didn't work for me. It just hung on SSH access.

This fork worked: https://github.com/unidevel/nixos-lima. Minor tweaks, mostly to use qcow2 format (reducing image size to 2.4GB vs >3GB). Not sure why it works and the original doesn't.

**Reply by unidevel** [reply]
Fixed by https://github.com/unidevel/nixos-lima/blame/main/lima-init.nix#L53, make `user-data` mounted correctly to make lima's ssh configuration working as expected.

**Comment by msgilligan** [comment]
Back in July I wrote:

>  I would potentially be interested in helping to make this a third-party add-on for Lima (since it is unlikely we'll be able to get it published by Nix.org or by the Lima maintainers.)

My nixos Lima VM that I installed with @patryk4815 's build is now broken -- DNS is not working in the guest. I'm not sure what broke it as I have upgraded both NixOS inside the guest as well as  Lima (using Nix Home Manager) on the host. I didn't notice the problem until I restarted the guest VM.

There are now at least 4 repositories listed above based on @patryk4815 's original work each with some user-specific customization and various fixes. I found a few more on GitHub. Some have added what look like useful features, but IMHO none of them looks like an ongoing project that other people can use and/or contribute to. Several look like they have made useful enhancements but IMHO none of them meet the criteria I am looking for to build a NixOS LimaVM that I could use for serious work. My criteria are:

1. Multiple contributors via PR 
2. Complete User documentation (how to install, customize, upgrade)
3. Issues tab is enabled and issues get responses
4. Developer documentation (explanation of how the modules work, how to contribute, etc)
5. CI (GitHub actions or other)

Now it's easy to be a critic. The people developing these modules clearly know much more about Nix than I _and_ they have made no promises of support and are clearly under no obligation. And I have limited bandwidth so I don't think I would be able to create neat another repository to solve the issues.

I'm just hoping that maybe a few of us can get together and combine our efforts to produce something the community can use and can be maintained long term. Looking at the individual efforts made on the various forks seems to indicate there is enough interest and talent to make this happen.

To this Nix novice, it seems like the best approach would be to create (one or more) flakes/modules that can be used to do the **core Lima-related magic necessary to emulate `cloud-init`**. This/these module(s) would need to work in two contexts:

1. In user-defined flakes used to build a Lima-bootable system image (e.g. `.qcow`)
2. In the `/etc/nixos` directory of the Lima guest VM so that the system can be updated with `nixos-rebuild`.

There could then be a separate template repo/flake that shows how to build a starter Lima-compatible image. Decoupling the module(s) from the starter/template repo would make it easier for people to customize to their hearts content, but be able to share the Lima modules and contribute upstream to improve them.

Does this make sense?  (I'm going to try tagging various people that have forked @patryk4815 's repo and see if we can get some discussion here...)

**Reply by msgilligan** [reply]
> at least 4 repositories

Here's a more complete list:

* @patryk4815  - https://github.com/patryk4815/ctftools/tree/master/lima-vm (the original!)
* @kasuboski - https://github.com/kasuboski/nixos-lima (inspired from patryk4815/ctftools)  Blog post: [NixOS Dev Environment on Mac](https://www.joshkasuboski.com/posts/nix-dev-environment) 
* @quinneden - https://github.com/quinneden/nixos-lima-vm (forked from kasuboski/nixos-lima)
* @unidevel - https://github.com/unidevel/nixos-lima (forked from kasuboski/nixos-lima)
* @nxmatic - https://github.com/nxmatic/nixos-lima (forked from kasuboski/nixos-lima, many interesting enhancements including `vz` support)

More can be found by looking at "forks" or "network" from kasuboski/nixos-lima, but the above are the ones that look most active to me. 

Who is interested in combining efforts and what do people think would be the best way to move forward?

**Reply by kradalby** [reply]
I want to pitch some ideas from https://github.com/cpick/nix-rosetta-builder too, it is initially planned as a replacement for `nix-darwin`'s linux-builder, but gave me some ideas:

- the config is fully declared as Nix, one could have a lima like wrapper for fully Nix defined VMs
  - VMs are managed as a service
  - Full rosetta support 
  - Integrated as build server
  - Could be automounted and replace colima as a docker endpoint for macOS
  - Multi VMs could be a thing, something that looks like https://github.com/astro/microvm.nix would be cool

**Reply by msgilligan** [reply]
> I want to pitch some ideas from https://github.com/cpick/nix-rosetta-builder too

That looks like a cool project and the features you list are very nice. I personally am interested in building VMs and managing them as services (i.e. via `launchd`) and am currently using the nix-darwin `linux-builder`. So I am definitely interested in seeing those ideas succeed.

However, I'm not sure how these ideas directly relate to what I am proposing here for Lima. I use Lima for creating long-lived VMs that I use for development and testing. The reasons I use Lima are that it is easy to spin up new VMs and to manage multiple VMs over time -- and the Lima feature set is very nice for that. My use of Lima predates use of Nix/NixOS. For me Lima is a familiar, practical tool that I used to manage Debian and Ubuntu VMs before wanting to add NixOS to the mix.

If the VMs are going to be built _and managed_ by Nix I don't see why Lima should be involved -- I think the pure Nix approach would be cool but at that point it has replaced the need for Lima. And I also see that some level of Lima/Colima compatibility might be useful in this Nix-based approach if you are using the VMs for running containers.

Are there features in `nix-rosetta-builder` that could help us build a flake that can builds a NixOS qcow image for Lima or help with `nixos-rebuild` within that image? (Besides the obvious case of using `nix-rossetta-builder` as a `linux-builder` replacement to run the flake-based image build)

**Reply by quinneden** [reply]
Just wanna include https://github.com/heywoodlh/flakes, the nixos-lima flake worked for me, based on the code it appears to be inspired by @kasuboski and @patryk4815.

**Reply by msgilligan** [reply]
Most of these seem to be broken with the latest release of Lima because `LIMA_CIDATA_COMMENT` was added to `lima.env` and this field typically contains a space.  The line:

```
        # ripped from https://github.com/lima-vm/alpine-lima/blob/main/lima-init.sh
        . "${LIMA_CIDATA_MNT}"/lima.env
```

seems to be the problem.  It was fixed upstream with: https://github.com/lima-vm/alpine-lima/commit/34b03daf6e5e55122e1e55b383e521fc5201a28d

**Reply by jandubois** [reply]
Yes, `lima.env` is in the same format as `/etc/environment`. It cannot be sourced directly by the shell, but needs to be read/processed. The same is true for `param.env` in the latest Lima releases.

**Reply by msgilligan** [reply]
Add https://github.com/charles-dyfis-net/nixos-lima to the list. @charles-dyfis-net has a series of useful commits from 2 days ago.

**Reply by msgilligan** [reply]
I have created a fork of @kasuboski 's repository, with several essential patches applied that allow building, running, and rebuilding (using the current `nixos-unstable` branch). I've created one-fix-per-commit changes in the `msgilligan-main` branch and updated the README to explain why I created yet-another fork and asking for collaborators/contributors.

https://github.com/msgilligan/nixos-lima

As I say in the README:

> If you have input or want to collaborate, please comment on Discussion #430 or open an issue or pull request on https://github.com/msgilligan/nixos-lima.

and:

> I'm also happy to archive this project and contribute to another one if other collaborators think that is a better path forward.

**Reply by msgilligan** [reply]
I created a new repository https://github.com/msgilligan/nixos-lima-config-sample that uses the `lima` module from the Flake in https://github.com/msgilligan/nixos-lima and allows you to manage the configuration of your NixOS Lima instance separately from the `nixos-lima` repository.

I hadn't seen this done before, but I believe @kasuboski created the `lima` module for this reason. This is my first time creating and sharing a NixOS module, so please forgive any rookie mistakes or sloppiness. It does seem to work for me, and I'm looking forward to receiving any feedback.


**UPDATE** there is now a GitHub organization and I've created a top-level comment to create a new thread about this:
https://github.com/lima-vm/lima/discussions/430#discussioncomment-12058172

**Comment by msgilligan** [comment]
I've created a [nixos-lima organization](https://github.com/nixos-lima/) based upon the work of @patryk4815 @kasuboski and others.

It contains a fork of @kasuboski 's repo that is not a GitHub fork: https://github.com/nixos-lima/nixos-lima

I have also split-out a "sample" of user-specific configuration that re-uses the `systemd` services defined in the main repository and can be used to customize an instance bootstrapped from the generated "generic" nixos-lima image: https://github.com/nixos-lima/nixos-lima-config-sample

I am hoping some of the creators of the other forks will be willing to submit patches as they all seem to know more about Nix than I do.

**Reply by msgilligan** [reply]
In **Issues** I've created a list of desired/possible enhancements: https://github.com/nixos-lima/nixos-lima/issues some of which exist on other forks.

@kasuboski @unidevel @patryk4815

**Reply by msgilligan** [reply]
It is based upon @kasuboski's last commit to `main`, but has the following additions:

* Fixes for latest Nixpkgs
* Fixes for latest Lima
* Added LICENSE file (MIT, same as Nixpkgs)
* Removal of user specific configuration
* Separate repository that demonstrates how to reference main flake and add user-specific configuration
* `services.lima.enable` configuration option to enable the module
* Fix to handle spaces in `lima.env`
* GitHub action workflow to build the system image
* GitHub action workflow that tests nixos-lima-config-sample
* Experimental Support for Lima's system and user provisioning scripts
* System image format changes to qcow format (for smaller size)

**Comment by msgilligan** [comment]
We are making progress on VZ support in [nixos-lima](https://github.com/nixos-lima/nixos-lima).

VZ support was literally Issue number 1 for `nixos-lima`  -- see  https://github.com/nixos-lima/nixos-lima/issues/1

This morning we received PR https://github.com/nixos-lima/nixos-lima/pull/10 with two changes (in 2 separate commits.) Both are needed to get VZ working. The first has been added to master, but the second (adding `--vsock-port 2222` to `lima-guestagent daemon`) fixes VZ while breaking QEMU, so we need to figure out  a mechanism for making it conditional upon the guest being run under VZ. (This should be decided at run-time of course, so the same `.qcow` image can be used with VZ or QEMU.)  Is there a recommended way to do that? 

**Update:** We found variables in `lima.env` that we can use: https://github.com/nixos-lima/nixos-lima/issues/1#issuecomment-3095021104

**Comment by msgilligan** [comment]
VZ support is merged to `master` on [nixos-lima](https://github.com/nixos-lima/nixos-lima)!

**Comment by msgilligan** [comment]
The following should now work:

```
limactl start https://raw.githubusercontent.com/nixos-lima/nixos-lima/master/nixos.yaml
```

This is experimental/pre-release, but I have been using this basic configuration as my daily-driver Lima NixOS VM on my aarch64 MacBook Pro for some time now.

**Reply by jandubois** [reply]
Thanks! Works for me...

I know this is your personal config; if you want to share it, I would prefer that the agent forwarding remains disabled, so users have to explicitly opt in, for security reasons.

I also don't understand the point of the provisioning scripts; they seem to be empty.

**Reply by msgilligan** [reply]
> I know this is your personal config;

It is not meant to be my personal config, see [Design Goals](https://github.com/nixos-lima/nixos-lima?tab=readme-ov-file#design-goals) -- number 3 and 4 are:

>3. User customization of NixOS Lima instance is done separately from initial image creation
> 4. Keep the base image and Nix services module as generic and reusable by others as possible


> I would prefer that the agent forwarding remains disabled, so users have to explicitly opt in, for security reasons.

I opened an issue to remind me to remove it: https://github.com/nixos-lima/nixos-lima/issues/18 -- **update** fixed/merged.

> I also don't understand the point of the provisioning scripts; they seem to be empty.

They are just a placeholder. I agree, they should be removed: https://github.com/nixos-lima/nixos-lima/issues/19 -- **update** fixed/merged.

**Comment by msgilligan** [comment]
Update: [nixos-lima](https://github.com/nixos-lima/nixos-lima) has had two more releases, usage has been simplified, and the README has been significantly improved.  (Note: the URL in https://github.com/lima-vm/lima/discussions/430#discussioncomment-13854295 has been updated/shortened)

If you're interested in using or supporting nixos-lima, you can help by:

* Visit the https://github.com/nixos-lima/nixos-lima project and "star" the repo.
* Use or test `nixos-lima` and provide feedback
* Publish a sample configuration or show how you use Nix deployment tools to deploy or rebuild your images
* Submit a PR with an improvement to the code
* Submit a PR with an improvement to the documentation

**Comment by msgilligan** [comment]
I have put a substantial amount of effort into having CI for the images, but am having trouble testing `aarch64` images on my free/public GitHub project. I just opened a separate "Discussion" for this issue to see if anyone has had any luck or has any suggestions: [How to test aarch64 templates / images on GitHub Actions #3934](https://github.com/lima-vm/lima/discussions/3934)

**Update:** I got it all working -- see https://github.com/lima-vm/lima/discussions/3934#discussioncomment-14261460

**Comment by msgilligan** [comment, marked as answer]
I have made a [v0.0.4 release](https://github.com/nixos-lima/nixos-lima/releases/tag/v0.0.4) of `nixos-lima` and it now uses the `nixos-25.11` stable branch and contains a few other fixes/improvements.

The repository is updated to support the new [GitHub Template URLs](https://lima-vm.io/docs/templates/github/) feature, so spinning up a new NixOS guest is as easy as:

```
limactl start github:nixos-lima
```

(Thanks to @jandubois and others for their work on the Template URL feature!)

**Comment by msgilligan** [comment]
[nixos-lima](https://github.com/nixos-lima/nixos-lima) release [v0.0.5](https://github.com/nixos-lima/nixos-lima/releases/tag/v0.0.5) is available.

## v0.0.5 Release Notes

### Released Images

* lima-init: support Lima 2.1+ with Instance ID -- [PR #52](https://github.com/nixos-lima/nixos-lima/pull/52) by @rinx
* Updated to `nixpkgs/nixos-25.11` - rev `54170c54449ea4d6725efd30d719c5e505f1c10e`

### Repository, documentation, and tests

* Support for starting a NixOS VM with `limactl start github:nixos-lima` -- see [Quickstart](https://github.com/nixos-lima/nixos-lima?tab=readme-ov-file#quickstart)
* `nixos.yaml`, `nixos-result.yaml`: configure `memory: 8GiB` -- fix for [Issue #48](https://github.com/nixos-lima/nixos-lima/issues/48)
* `nixos.yaml`, `nixos-result.yaml`: port forward ignore rule for DHCP client port  -- [PR #51](https://github.com/nixos-lima/nixos-lima/pull/51) fixes [Issue #48](https://github.com/nixos-lima/nixos-lima/issues/48)
* Use `nixpkgs-unstable` for lima 2.0.3 in devShell, GitHub Actions -- [PR #58](https://github.com/nixos-lima/nixos-lima/pull/58)
* Use a Nix formatter for the `.nix` files -- [PR #57](https://github.com/nixos-lima/nixos-lima/pull/57)
* Updated and improved GitHub Actions

### Contributors

@rinx, @justin-qu
